third_party/move/move-prover/bytecode-pipeline/src/memory_instrumentation.rs

```diff
@@ -59,8 +66,114 @@ impl FunctionTargetProcessor for MemoryInstrumentationProcessor {
+fn process_spec_clauses(
+    code: &[Bytecode],
+    offset: &mut usize,
+    func_env: &FunctionEnv,
+    instrumenter: &mut Instrumenter<'_>,
+) {
+    let (first_attr_id, _, exp) = match &code[*offset] {
+        Prop(attr_id, _, exp) => (attr_id, /* ignored */ (), exp),
+        _ => unreachable!("Expected a Prop"),
+    };
+
+    // Get the borrow info before the first spec clause.
+    let borrow_annotation_at = instrumenter
+        .borrow_annotation
+        .get_borrow_info_at(*offset as CodeOffset)
+        .unwrap();
+    let borrow_info = &borrow_annotation_at.before;
+
+    // Process the first clause.
+    let mut spec_instrs = vec![code[*offset].clone()];
+    let mut nodes = Vec::new();
+    collect_borrow_nodes_from_exp(&mut nodes, func_env, borrow_info, exp);
+    *offset += 1;
+
+    // Process all consecutive ones
+    while *offset < code.len() {
+        if let Prop(_, _, exp) = &code[*offset] {
+            collect_borrow_nodes_from_exp(&mut nodes, func_env, borrow_info, exp);
+            spec_instrs.push(code[*offset].clone());
+            *offset += 1;
+        } else {
+            break;
+        }
+    }
+
+    // Determine whether loop invariants
+    let loop_invariant_attr = if instrumenter
+        .builder
+        .data
+        .loop_invariants
+        .contains(first_attr_id)
+    {
+        Some(first_attr_id)
+    } else {
+        None
+    };
+
+    // Either record write-back actions for loop invariants, or instrument them directly
+    match loop_invariant_attr {
+        Some(attr)
+            if !instrumenter
+                .builder
+                .data
+                .loop_invariant_write_back_map
+                .contains_key(attr) =>
+        {
+            instrumenter
+                .builder
+                .data
+                .loop_invariant_write_back_map
+                .insert(*attr, (borrow_info.clone(), nodes));
+        },
+        _ => {
+            Instrumenter::instrument_write_back_for_spec(
+                &mut instrumenter.builder,
+                borrow_info,
+                nodes,
+            );
+        },
+    }
+
+    // Emit all the spec instructions.
+    for instr in spec_instrs {
+        instrumenter.builder.emit(instr);
+    }
+}
+
+fn collect_borrow_nodes_from_exp(
+    nodes: &mut Vec<BorrowNode>,
+    func_env: &FunctionEnv,
+    borrow_info: &BorrowInfo,
+    exp: &Exp,
+) {
+    let env = func_env.module_env.env;
+    exp.used_temporaries_with_types(env)
+        .iter()
+        .map(|(temp, e)| {
+            if e.is_reference() {
+                BorrowNode::Reference(*temp)
+            } else {
+                BorrowNode::LocalRoot(*temp)
+            }
+        })
+        .chain(
+            exp.used_memory(env)
+                .iter()
+                .map(|(id, _)| BorrowNode::GlobalRoot(id.clone())),
+        )
+        .filter(|node| borrow_info.has_borrow(node))
+        .for_each(|node| {
+            if !nodes.contains(&node) {
```

Comment:
Should `nodes` be a set instead of a vector, if we want to maintain uniqueness?

Reply:
done